### PR TITLE
git: add defaultBranchName

### PR DESCRIPTION
**Pull request by rhubert**
Fixes: #429

Not sure if this fits. The higher layer would override the setting of a lower layer... :/

### CODE REVIEW

**Comment by jkloetzke** [comment]
> Not sure if this fits. The higher layer would override the setting of a lower layer... :/

I think it's the other way around. A sub-layer could override the setting of the upper layer. This is certainly not a good idea.

Given that we have `scmOverrides` in `default.yaml` it feels more natural to move this setting there. And while we're at it, we might add the possibility to change other defaults too. Maybe something like:

```yaml
scmDefaults:
    git:
        branch: main
        singleBranch: True
```

What do you think?

File: doc/manual/configuration.rst [hunk 1]
```diff
@@ -876,31 +876,31 @@ cvs    CVS repository               | ``cvsroot``: repository location ("``:ext:
                                     | ``module``: module name
                                     | ``rev``: revision, branch, or tag name (optional)
 git    `Git`_ project               | ``url``: URL of remote repository
-                                    | ``branch``: Branch to check out (optional, default: master)
+                                    | ``branch`` (*): Branch to check out (optional, default: master)
```
Comment (jkloetzke) [inline doc/manual/configuration.rst:879]: The single asterisk must be escaped => `(\*)`

File: doc/manual/configuration.rst [hunk 2]
```diff
@@ -2079,6 +2079,14 @@ the anonymous access to the container can be used like this::
...
 in case other backends are configured too.
 
+scmDefaults
+~~~~~~~~~~~
+
+Type: Dict of SCM dicts
+
+Default settings for SCMs are applied if the value is not set in the
+recipe. Useable values are marked with `(*)` in (:ref:`configuration-recipes-scm`).
+
```
Comment (jkloetzke) [inline doc/manual/configuration.rst:2089]: It would be good to have some example to see how the syntax looks exactly.

File: pym/bob/input.py [hunk 3]
```diff
@@ -389,6 +389,13 @@ def Scm(spec, env, overrides, recipeSet):
         except schema.SchemaError as e:
             raise ParseError("Error validating SCM after applying scmOverrides: {}".format(str(e)))
 
+    # apply scmDefaults
+    defaults = recipeSet.scmDefaults().get(spec['scm'])
+    if defaults:
+        for d in defaults:
+            if not spec.get(d):
+                spec[d] = defaults[d]
```
Comment (jkloetzke) [inline pym/bob/input.py:397]: I think this is equivalent to:

```suggestion
    for k, v in recipeSet.scmDefaults().get(spec['scm'], {}).items():
        spec.setdefault(k, v)
```

I guess we'll also need some test that uses the feature.

**Review by jkloetzke — COMMENTED** [review]
This looks quite good IMO. Just some minor comments...

File: pym/bob/input.py (hunk 3 above)
Comment (rhubert) [reply pym/bob/input.py]: It's always amazing ... :+1:

File: doc/manual/configuration.rst [hunk 4]
```diff
@@ -852,7 +852,7 @@ If the package is built from multiple modules you can give a list of SCMs::
             dir: src/bar
 
 There are three common (string) attributes in all SCM specifications: ``scm``,
-``dir`` and ``if``. Additionally to the string type, the ``if`` property may
+``dir`` (\*) and ``if``. Additionally to the string type, the ``if`` property may
```
Comment (jkloetzke) [inline doc/manual/configuration.rst:855]: Looks like the `cvs` SCM still missing this change...

File: pym/bob/input.py (hunk 3 above)
Comment (jkloetzke) [reply pym/bob/input.py]: :grinning:

File: doc/manual/configuration.rst (hunk 4 above)
Comment (rhubert) [reply doc/manual/configuration.rst:855]: you are right.. The unit test will also follow..

File: test/test_input_recipeset.py [hunk 5]
```diff
@@ -1390,3 +1390,171 @@ def testNewBehaviour(self):
 
         pkg = self.generate().walkPackagePath("root")
         self.assertTrue(pkg.getCheckoutStep().getScmList()[0].getProperties(False)["prune"])
+
+
+class TestScmDefaults(RecipesTmp, TestCase):
+    """ Test behaviour of pruneImportScm policy"""
```
Comment (jkloetzke) [inline test/test_input_recipeset.py:1396]: Copy'n'paste

File: test/test_input_recipeset.py [hunk 6]
```diff
@@ -1390,3 +1390,171 @@ def testNewBehaviour(self):
...
+
+    def testScmDefaultsGit(self):
+        """Test scmDefaults for gitscm"""
+        with TemporaryDirectory() as tmp:
+            os.chdir(tmp)
+            os.mkdir("recipes")
```
Comment (jkloetzke) [inline test/test_input_recipeset.py:1402]: No need to do this. The `RecipesTmp` base class does exactly this already.

File: test/test_input_recipeset.py [hunk 7]
```diff
@@ -1390,3 +1390,171 @@ def testNewBehaviour(self):
...
+                   scm: git
+                   url: foo@bar
+                buildScript: "true"
+                packageScript: "true"
+                """)
+            pkg = self.generate().walkPackagePath("root")
+            self.assertEqual(pkg.getCheckoutStep().getScmList()[0].getProperties(False)["branch"],
+                    "master")
+            self.assertEqual(pkg.getCheckoutStep().getScmList()[0].getProperties(False)["sslVerify"],
+                    None)
```
Comment (jkloetzke) [inline test/test_input_recipeset.py:1416]: How about this to reduce the redundancy?

```suggestion
            props = pkg.getCheckoutStep().getScmList()[0].getProperties(False)
            self.assertEqual(props["branch"], "master")
            self.assertEqual(props["sslVerify"], None)
```
...

File: test/test_input_recipeset.py [hunk 8]
```diff
@@ -1390,3 +1390,171 @@ def testNewBehaviour(self):
...
+            self.assertEqual(pkg.getCheckoutStep().getScmList()[0].getProperties(False)["shallow"],
+                    None)
+            self.assertEqual(pkg.getCheckoutStep().getScmList()[0].getProperties(False)["singleBranch"],
+                    None)
+            self.assertFalse(pkg.getCheckoutStep().getScmList()[0].getProperties(False)["submodules"])
+            self.assertFalse(pkg.getCheckoutStep().getScmList()[0].getProperties(False)["recurseSubmodules"])
+            self.assertTrue(pkg.getCheckoutStep().getScmList()[0].getProperties(False)["shallowSubmodules"])
+
+            with open("default.yaml", "w") as f:
```
Comment (jkloetzke) [inline test/test_input_recipeset.py:1425]: Maybe you could add a `writeDefault` (like the existing `writeConfig`) helper in the `RecipesTmp` class?

**Comment by jkloetzke** [comment]
Thanks :+1:

**Pull request merged**